### PR TITLE
CC13XX CC26XX : transition from pinmux to pinctrl

### PR DESCRIPTION
**Pull request by vaishnavachath**
This pull request contains the changes to migrate from pinmux to pinctrl API for CC13XX and CC26XX based platforms as outlined in https://github.com/zephyrproject-rtos/zephyr/issues/39740 .
The pull request contains the following changes in the respective commits:

1.  Introduces the pinctrl driver for CC13XX/CC26XX platform.
2. migrate existing consumers to use the new driver, pinctrl API.
3. remove obsolete pinmux driver.
4. fix white-space/styling issues for defines moved from hal/driverlib/ioc.h - intentionally kept as separate commit for now, can be squashed with the first commit  if that is preferred.

The change-set was tested on CC1352R sensor-tag to verify the I2C/SPI/UART IOC functionality is not broken.

### CODE REVIEW

**Comment by vaishnavachath** [comment]
@cfriedt @bwitherspoon Can you please review the PR?

**Review by cfriedt — DISMISSED** [review]
LGTM - thanks so much for getting these changes in!

@vaishnavachath - would be great if you could convince TI to allow you to be platform maintainer for Zephyr 😃

File: drivers/pinctrl/Kconfig.cc13xx_cc26xx [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+# SPDX-License-Identifier: Apache-2.0
+
+DT_COMPAT_CC13XX_CC26XX_PINCTRL := ti,cc13xx-cc26xx-pinctrl
+
+config PINCTRL_CC13XX_CC26XX
+	bool "TI SimpleLink CC13xx / CC26xx pinctrl driver"
+	depends on SOC_SERIES_CC13X2_CC26X2
+	default $(dt_compat_enabled,$(DT_COMPAT_CC13XX_CC26XX_PINCTRL))
+	help
+	  Enable the TI SimpleLink CC13xx / CC26xx pinctrl driver
```
Comment (gmarull) [inline drivers/pinctrl/Kconfig.cc13xx_cc26xx]: nit: missing newline at the eof

File: dts/bindings/pinctrl/ti,cc13xx-cc26xx-pinctrl.yaml [hunk 2]
```diff
@@ -0,0 +1,28 @@
+# Copyright (c) 2022 Vaishnav Achath
+# SPDX-License-Identifier: Apache-2.0
+
+description: |
+    TI SimpleLink CC13xx / CC26xx pinctrl node.
+
+    Device pin configuration should be placed in the child nodes of this node.
+    Populate the 'pinmux' field with a pair consisting of a pin number and its IO
+    function
```
Comment (gmarull) [inline dts/bindings/pinctrl/ti,cc13xx-cc26xx-pinctrl.yaml:9]: please extend this documentation with some examples, supported properties, etc. See other vendors as an example.

File: include/dt-bindings/pinctrl/cc13xx_cc26xx-pinctrl.h [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+#ifndef CC13XX_CC26XX_PINCTRL_COMMON_H_
+#define CC13XX_CC26XX_PINCTRL_COMMON_H_
+
+/* Adapted from hal/ti/simplelink/source/ti/devices/cc13x2_cc26x2/driverlib/ioc.h */
+
+//*****************************************************************************
+//
+// IOC Peripheral Port Mapping
+//
+//*****************************************************************************
```
Comment (gmarull) [inline include/dt-bindings/pinctrl/cc13xx_cc26xx-pinctrl.h]: please no C99 style comments

File: include/dt-bindings/pinctrl/cc13xx_cc26xx-pinctrl.h [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+#define IOC_STD_OUTPUT          (IOC_CURRENT_2MA | IOC_STRENGTH_AUTO |      \
+                                 IOC_NO_IOPULL | IOC_SLEW_DISABLE |         \
+                                 IOC_HYST_DISABLE | IOC_NO_EDGE |           \
+                                 IOC_INT_DISABLE | IOC_IOMODE_NORMAL |      \
+                                 IOC_NO_WAKE_UP | IOC_INPUT_DISABLE )
+ #define IOC_STD_I2C            (IOC_CURRENT_2MA | IOC_STRENGTH_AUTO |              \
+                                 IOC_IOPULL_UP | IOC_SLEW_DISABLE |                 \
+                                 IOC_HYST_DISABLE | IOC_NO_EDGE |                   \
+                                 IOC_INT_DISABLE | IOC_IOMODE_OPEN_DRAIN_NORMAL |   \
+                                 IOC_NO_WAKE_UP | IOC_INPUT_ENABLE )
```
Comment (gmarull) [inline include/dt-bindings/pinctrl/cc13xx_cc26xx-pinctrl.h]: standard properties need to be used. For example, `bias-pull-up`. This is encoding pin properties (bias, pull-up, slew-rate, etc.) in a custom way. Ref. https://docs.zephyrproject.org/latest/guides/pinctrl/index.html#pin-control-drivers

File: boards/arm/cc1352r1_launchxl/cc1352r1_launchxl-pinctrl.dtsi [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <dt-bindings/pinctrl/cc13xx_cc26xx-pinctrl.h>
+
+&pinctrl {
+    /* UART0 */
+    uart0_tx_default: uart0_tx_default {
+		pinmux = <13 IOC_PORT_MCU_UART0_TX IOC_STD_OUTPUT>;
```
Comment (gmarull) [inline boards/arm/cc1352r1_launchxl/cc1352r1_launchxl-pinctrl.dtsi:12]: nit: please don't mix tabs/spaces

**Review by gmarull — CHANGES_REQUESTED** [review]

File: include/dt-bindings/pinctrl/cc13xx_cc26xx-pinctrl.h (hunk 3 above)
Comment (vaishnavachath) [reply include/dt-bindings/pinctrl/cc13xx_cc26xx-pinctrl.h]: this file is a trimmed down version from hal/ti/..driverlib/ioc.h , all styling issues fixed for this file as separate commit

**Review by mbolivar-nordic — COMMENTED** [review]
Other than @gmarull's comment on missing examples and details, the bindings changes LGTM.

**Comment by cfriedt** [comment]
@vaishnavachath - can you make the changes requested by @gmarull ?

File: dts/bindings/pinctrl/ti,cc13xx-cc26xx-pinctrl.yaml (hunk 2 above)
Comment (vaishnavachath) [reply dts/bindings/pinctrl/ti,cc13xx-cc26xx-pinctrl.yaml]: added documentation

File: include/dt-bindings/pinctrl/cc13xx_cc26xx-pinctrl.h (hunk 4 above)
Comment (vaishnavachath) [reply include/dt-bindings/pinctrl/cc13xx_cc26xx-pinctrl.h]: implemented standard properties

**Comment by vaishnavachath** [comment]
@gmarull @cfriedt @mbolivar-nordic , I have made the suggested changes, can you please review the updates.
Changes made:
 - implement standard properties (bias-pull-up, bias-pull-down .etc)
 - update DT binding documentation with examples
 - earlier the styling fixes on driverlib header was separate commit which created confusion, that was fixed in the initial commit itself

**Review by mbolivar-nordic — DISMISSED** [review]
Bindings LGTM. The rest I only reviewed quickly, but it all seems reasonable as well. Thanks for the update!

**Comment by cfriedt** [comment]
@vaishnavachath - looks like some minor changes to get the last two tests to pass. Are you running with Twister?

File: drivers/pinctrl/pinctrl_cc13xx_cc26xx.c [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+
+	return 0;
+}
+
+int pinctrl_configure_pins(const pinctrl_soc_pin_t *pins, uint8_t pin_cnt, uintptr_t reg)
+{
+	ARG_UNUSED(reg);
+	int i;
+
+	for (i = 0; i < pin_cnt; i++) {
```
Comment (gmarull) [inline drivers/pinctrl/pinctrl_cc13xx_cc26xx.c:29]:
```suggestion
	for (uint8_t i = 0U; i < pin_cnt; i++) {
```

**Review by gmarull — DISMISSED** [review]
LGTM, one minor non-blocking nitpick

File: drivers/pinctrl/pinctrl_cc13xx_cc26xx.c (hunk 6 above)
Comment (vaishnavachath) [reply drivers/pinctrl/pinctrl_cc13xx_cc26xx.c]: Updated with the suggested change

**Comment by vaishnavachath** [comment]
> @vaishnavachath - looks like some minor changes to get the last two tests to pass. Are you running with Twister?

@cfriedt made small mistake in cc26xx dtsi while updating which is fixed now and tests are passing.
@gmarull updated with suggested minor change in `pinctrl_cc13xx_cc26xx.c`

**Review by cfriedt — APPROVED** [review]

**Review by gmarull — APPROVED** [review]

**Pull request merged**